url74c52881d6_AndrewBourgeois_JGWCAConstants_tgz-pJ8-be_demmel_fun_jgwcaconstants_Tutorial4J8/src/be/demmel/fun/jgwcaconstants/bll/GWCAOperations.java

```diff
@@ -0,0 +1,2819 @@
+        return SkillType.toEnum((short) byteArrayToInt(lParam));
+    }
+
+    /**
+     * @param bag
+     * @param itemSlot
+     * @throws IOException 
+     */
+    public void prepareMoveItem(BagLocation bag, int itemSlot) throws IOException {
+        sendPacket(GWCAOperation.PREPARE_MOVE_ITEM, bag.getValue(), itemSlot);
+    }
+
+    /**
+     * @param itemId
+     * @throws IOException 
+     */
+    public void prepareMoveItem(int itemId) throws IOException {
+        sendPacket(GWCAOperation.MOVE_ITEM, itemId);
+    }
+
+    /**
+     * @param bag
+     * @param itemSlot
+     * @throws IOException 
+     */
+    public void moveItem(BagLocation bag, int itemSlot) throws IOException {
+        sendPacket(GWCAOperation.MOVE_ITEM, bag.getValue(), itemSlot);
+    }
+
+    /**
+     * Moves an item from one location to another
+     * @param currentBag
+     * @param currentItemSlot
+     * @param targetBag
+     * @param targetItemSlot
+     * @throws IOException 
+     */
+    public void moveItem(BagLocation currentBag, int currentItemSlot, BagLocation targetBag, int targetItemSlot) throws IOException {
+        this.prepareMoveItem(currentBag, currentItemSlot);
+        try {
+            //TODO: explain why this sleep is needed
+            Thread.sleep(20);
+        } catch (InterruptedException ex) {
+            // ignore
+        }
+
+        this.moveItem(targetBag, targetItemSlot);
+    }
+
+    /**
+     * Moves an item from one location to another
+     * @param itemId
+     * @param targetBag
+     * @param targetItemSlot
+     * @throws IOException 
+     */
+    public void moveItemById(int itemId, BagLocation targetBag, int targetItemSlot) throws IOException {
+        this.prepareMoveItem(itemId);
+        try {
+            //TODO: explain why this sleep is needed
+            Thread.sleep(20);
+        } catch (InterruptedException ex) {
+            // ignore
+        }
+
+        this.moveItem(targetBag, targetItemSlot);
+    }
+
+    /**
+     * @param numberOfItems
+     * @param maxDistance
+     * @throws IOException 
+     */
+    public void pickupItems(int numberOfItems, float maxDistance) throws IOException {
+        int itemsPicked = 0;
+
+        long startDeadlockCheck = System.currentTimeMillis();
+
+        do {
+            float[] itemAgentIdAndDistance = this.getNearestItemToAgentEx(-2);
+
+            long currentlyElapsedTime = System.currentTimeMillis() - startDeadlockCheck;
+            if (itemAgentIdAndDistance[0] == 0 || itemAgentIdAndDistance[1] > maxDistance || currentlyElapsedTime > 30000) {
+                break;
+            }
+
+            boolean itemStillExists = true;
+            do {
+                //FIXME: why pickup item twice??
+                this.pickupItem((int) itemAgentIdAndDistance[0]);
+                try {
+                    //FIXMR: why is there a sleep needed?
+                    Thread.sleep(500);
+                } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)